### DISCUSSION TITLE
Handle undefined field

### DISCUSSION THREAD

**Discussion by bigbean17** [emqx/emqx#13608] [Q&A] [answered]
Hi,

We are trying to integrate emqx with Influxdb. An example mqtt message is
```
{
    data: {
        device_a: {value, id},
        device_b:{value, id},
        ....
}
```
Some device might be missing from the mqtt message, which indicates that device is not currently connected. However, the SQL will give output looks like below if device_b is not in the mqtt message. 
```
{
    device_a_value: 10.3,
    device_b_value: "undefined",
    ...
```
}
that "undefined" value causes the insertion to influxdb fail. We tried to use `case` clause to set the value to null but it still show undefined. Is there any way to solve this problem? The values are floats. 

Thanks for helping

**Comment by zmstone** [comment]
Hi @bigbean17 

this should work.
```
SELECT
  payload.data.device_a.value as device_a_value,
  CASE WHEN is_null(payload.data.device_b.value) THEN -1 ELSE payload.data.device_b.value END as device_b_value
FROM
  "t/#"
```

**Reply by bigbean17** [reply]
Hi,

Thanks for the reply, but this case it only works when the key is missing from the message. For example, my message is
```
{
    data: {"x": 123}
}
```
But you are selecting "t", in this case, `is_null(payload.data.t.value)` will return true. However, if my message is
```
{
    data: {"x": 123, "t": null}
}
```
And you select "t", in this case, `is_null(payload.data.t.value)` will return false, even though the output value of t in the SQL test is still `null`.

**Reply by thalesmg** [reply]
Hi.  Did you try the rule and it returned false?

`is_null` should return true if intermediate fields in the path are missing as well.

e.g.:

```sql
SELECT
  is_null(payload.t.field) as is_field_null
FROM
  "t/#"
```
with input `{"msg": "hello"}` yields `is_field_null: true`.

![image](https://github.com/user-attachments/assets/62eddbdd-12a0-4a13-9c14-5c5e41ef60b5)

**Comment by zmstone** [comment, marked as answer]
Hi @bigbean17 
JSON null is treated as value in EMQX rules.
Here is how it should work:

```
SELECT
  CASE
    WHEN is_null(payload.a) THEN
      'IS-MISSING'
    WHEN payload.a = 'null' THEN
      'IS-NULL'
    ELSE
      'NOT-NULL'
  END AS is_a_null
FROM
  "t/#"
```

**Reply by zmstone** [reply]
there is actually an undocumented function `is_null_var` which checks if it's a JSON `null`.
will add it to the doc.

**Reply by bigbean17** [reply]
cool, that will be handy, thank you!
